packages/@vue/cli-shared-utils/lib/env.js

```diff
@@ -101,10 +96,32 @@ exports.hasPnpm3OrLater = () => {
     // so we only support pnpm >= 3.0.0
     _hasPnpm = true
     _hasPnpm3orLater = semver.gte(pnpmVersion, '3.0.0')
-    return _hasPnpm3orLater
+    _hasPnpm4orLater = semver.gte(pnpmVersion, '4.0.0')
   } catch (e) {
-    return (_hasPnpm3orLater = false)
+    _hasPnpm3orLater = false
+    _hasPnpm4orLater = false
+  }
+  return [_hasPnpm3orLater, _hasPnpm4orLater]
+}
+
+exports.hasPnpm3OrLater = () => {
+  if (process.env.VUE_CLI_TEST) {
+    return true
+  }
+  if (_hasPnpm3orLater != null) {
+    return _hasPnpm3orLater
+  }
+  return checkPnpmVersion()[0]
+}
+
+exports.hasPnpm4OrLater = () => {
+  if (process.env.VUE_CLI_TEST) {
+    return true
+  }
+  if (_hasPnpm4orLater != null) {
+    return _hasPnpm4orLater
   }
+  return checkPnpmVersion()[1]
```

Comment:
```suggestion
  checkPnpmVersion()
  return _hasPnpm4orLater
```

Reply:
I am not really sure if I now can also remove the return value from checkPnpmVersion(). Will it be used for testing?

Reply:
@sodatea What do you think?

Reply:
Why not just cache the PNPM version and move the semver logic into the two exported functions?
```js
exports.hasPnpm4OrLater = () => {
  // ...
  return semver.gte(getPnpmVersion() || '0.0.0', '4.0.0')
}
```

Reply:
Currently I don't see a benefit of caching the version number. Wouldn't that introduce more function calls or am I overseeing something?  What do you think?

Reply:
1. While you name the function as `checkPnpmVersion`, it only does the job of `checkPnpmVersion3or4`. What if we want to add a `hasPnpm5OrLater` function? The current implementation is a little bit obscure and not very extensible in the long run.
2. The cost of a normal function call is negligible in most cases. What's not negligible is the cost of forking a process, namely the `execSync` call, which may take a few hundred milliseconds to finish. So we need to avoid calling `execSync` multiple times, but no need to worry about other functions.

Reply:
Indeed good points. I have implemented the caching now. I have also removed the two functions `hasPnpm3/4OrLater` which can be now replicated by using the `hasPnpmXOrLater(version)` function.